src/addrman.h

```diff
@@ -318,8 +329,15 @@ friend class CAddrManTest;
     {
         LOCK(cs);
 
-        unsigned char nVersion = 2;
-        s << nVersion;
+        const auto stream_version_orig = s.GetVersion();
+
+        if (m_format >= Format::V3) {
+            // Add ADDRV2_FORMAT to the version so that the CNetAddr and CAddress
+            // serialize methods produce an address in v2 format.
+            s.SetVersion(stream_version_orig | ADDRV2_FORMAT);
```

Comment:
A cleaner approach is using `OverrideStream` here, which removes the need to undo the version change, or having a mutable version in CHashWriter.

Reply:
Right, changed to `OverrideStream`.

https://github.com/bitcoin/bitcoin/pull/19503 is also an alternative if/when it gets merged.